### PR TITLE
[Prover] Fix bug in loop unrolling

### LINKED ISSUE #15044: [Bug] [Move Prover] CFG simplification breaks up the prover
# 🐛 Bug

Running
```
MOVE_COMPILER_V2=1 cargo test --package aptos-framework --test move_prover_tests -- move_aptos_stdlib_prover_tests --exact --show-output
```

gives

```
running 1 test
error: post-condition does not hold
   ┌─ /Users/zw/aptos-core/aptos-move/framework/aptos-stdlib/sources/comparator.spec.move:40:9
   │
40 │ ╭         ensures (result.inner == EQUAL) ==> (
41 │ │             (left_length == right_length) &&
42 │ │                 (forall i: u64 where i < left_length: left[i] == right[i])
43 │ │         );
   │ ╰──────────^
   │
   =     at /Users/zw/aptos-core/aptos-move/framework/aptos-stdlib/sources/comparator.move:38: compare_u8_vector
   =     at /Users/zw/aptos-core/aptos-move/framework/aptos-stdlib/sources/comparator.spec.move:37
   =     at /Users/zw/aptos-core/aptos-move/framework/aptos-stdlib/sources/comparator.spec.move:38
   =         left = vector{190u8, 34u8}
   =         right = vector{190u8, 33u8}
   =     at /Users/zw/aptos-core/aptos-move/framework/aptos-stdlib/sources/comparator.move:39: compare_u8_vector
   =         left_length = 2
   =     at /Users/zw/aptos-core/aptos-move/framework/aptos-stdlib/sources/comparator.move:40: compare_u8_vector
   =         right_length = 2
   =     at /Users/zw/aptos-core/aptos-move/framework/aptos-stdlib/sources/comparator.move:42: compare_u8_vector
   =         idx = 0
   =     at /Users/zw/aptos-core/aptos-move/framework/aptos-stdlib/sources/comparator.move:44: compare_u8_vector
   =         $t9 = true
   =     at /Users/zw/aptos-core/aptos-move/framework/aptos-stdlib/sources/comparator.move:44: compare_u8_vector
   =     at /Users/zw/aptos-core/aptos-move/framework/aptos-stdlib/sources/comparator.move:45: compare_u8_vector
   =         left_byte = 190u8
   =     at /Users/zw/aptos-core/aptos-move/framework/aptos-stdlib/sources/comparator.move:46: compare_u8_vector
   =         right_byte = 190u8
   =     at /Users/zw/aptos-core/aptos-move/framework/aptos-stdlib/sources/comparator.move:48: compare_u8_vector
   =     at /Users/zw/aptos-core/aptos-move/framework/aptos-stdlib/sources/comparator.move:50: compare_u8_vector
   =     at /Users/zw/aptos-core/aptos-move/framework/aptos-stdlib/sources/comparator.move:53: compare_u8_vector
   =         idx = 1
   =     at /Users/zw/aptos-core/aptos-move/framework/aptos-stdlib/sources/comparator.move:44: compare_u8_vector
   =         $t9 = true
   =     at /Users/zw/aptos-core/aptos-move/framework/aptos-stdlib/sources/comparator.move:61: compare_u8_vector
   =         $t9 = false
   =     at /Users/zw/aptos-core/aptos-move/framework/aptos-stdlib/sources/comparator.move:44: compare_u8_vector
   =     at /Users/zw/aptos-core/aptos-move/framework/aptos-stdlib/sources/comparator.move:56: compare_u8_vector
   =     at /Users/zw/aptos-core/aptos-move/framework/aptos-stdlib/sources/comparator.move:58: compare_u8_vector
   =     at /Users/zw/aptos-core/aptos-move/framework/aptos-stdlib/sources/comparator.move:59: compare_u8_vector
   =     at /Users/zw/aptos-core/aptos-move/framework/aptos-stdlib/sources/comparator.spec.move:3
   =     at /Users/zw/aptos-core/aptos-move/framework/aptos-stdlib/sources/comparator.move:61: compare_u8_vector
   =         return = comparator.Result{inner = 0u8}
   =     at /Users/zw/aptos-core/aptos-move/framework/aptos-stdlib/sources/comparator.move:57: compare_u8_vector
   =         result = comparator.Result{inner = 0u8}
   =     at /Users/zw/aptos-core/aptos-move/framework/aptos-stdlib/sources/comparator.spec.move:35
   =     at /Users/zw/aptos-core/aptos-move/framework/aptos-stdlib/sources/comparator.spec.move:40

test move_aptos_stdlib_prover_tests ... FAILED
```

even though the compiled bytecode is still correct for `compare_u8_vector`.

## Expected Behavior

The test should pass.

### PR DESCRIPTION
**Pull request by rahxephon89**
## Description

Currently, loop analysis [collects](https://github.com/aptos-labs/aptos-core/blob/3b23a90154809474c905d8e90c6bd0d4b8f7ef0f/third_party/move/move-model/bytecode/src/fat_loop.rs#L447) bytecode for unrolling loops and returns a vector of basic blocks, which will be directly flatten later. This implementation will lead to incorrect boogie code if a block in the vector falls through to its successor in the original CFG but the next element in the vector is actually not its successor. The bug leads to verification failure shown in #15044 when CFG simplification is turned on. This PR fixes the issue by inserting the explicit jump instruction.

Close #15044 

## How Has This Been Tested?

1) existing tests pass;
2) a new test;

## Key Areas to Review

The code change will only influence bounded verification with `unroll` pragma. Need to make sure this change will not affect other part of the prover.

## Type of Change
- [ ] New feature
- [x] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [ ] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Move Compiler
- [x] Prover

## Checklist
- [x] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I identified and added all stakeholders and component owners affected by this change as reviewers
- [x] I tested both happy and unhappy path of the functionality
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

File: third_party/move/move-prover/tests/sources/functional/bv_aborts.v2_exp [hunk 1]
```diff
@@ -8,22 +8,24 @@ error: function does not abort under this condition
    =     at tests/sources/functional/bv_aborts.move:11: assert_with_spec
    =         x = <redacted>
    =     at tests/sources/functional/bv_aborts.move:12: assert_with_spec
-   =     at tests/sources/functional/bv_aborts.move:13: assert_with_spec
    =     at tests/sources/functional/bv_aborts.move:16: assert_with_spec (spec)
 
 error: abort not covered by any of the `aborts_if` clauses
    ┌─ tests/sources/functional/bv_aborts.move:14:5
    │
-12 │           assert!(x > 815);
```
Comment (rahxephon89) [inline third_party/move/move-prover/tests/sources/functional/bv_aborts.v2_exp:17]: This is due to cfg simplification

File: third_party/move/move-model/bytecode/ast-generator-tests/tests/conditionals.exp [hunk 2]
```diff
@@ -20,11 +20,11 @@ fun m::if_1($t0|c: bool): u8 {
   1: $t1 := $t2
   2: $t3 := move($t0)
   3: if ($t3) goto 4 else goto 8
-  4: label L1
```
Comment (rahxephon89) [inline third_party/move/move-model/bytecode/ast-generator-tests/tests/conditionals.exp:23]: change of exp files in this crate is due to cfg simplification.

File: third_party/move/tools/move-decompiler/tests/fixed_point32.exp [hunk 3]
```diff
@@ -15,24 +15,23 @@ module 0x1::fixed_point32 {
         *&(&self).value >> 32u8
     }
     public fun create_from_rational(numerator: u64, denominator: u64): FixedPoint32 {
-        let _t5;
```
Comment (rahxephon89) [inline third_party/move/tools/move-decompiler/tests/fixed_point32.exp:18]: change of exp files in this crate is due to cfg simplification.

**Review by brmataptos — APPROVED** [review]
PR looks ok.  Existing tests and error outputs look like they could be improved.

**Review by fEst1ck — APPROVED** [review]
LGTM

File: third_party/move/move-model/bytecode/ast-generator-tests/tests/abort_example.exp [hunk 4]
```diff
@@ -18,58 +18,50 @@ public fun m::f($t0|length: u64): u64 {
      var $t5: bool
      var $t6: u64
      var $t7: u64
-     var $t8: u64
```
Comment (rahxephon89) [inline third_party/move/move-model/bytecode/ast-generator-tests/tests/abort_example.exp:21]: change of exp files in this crate is due to cfg simplification.

File: third_party/move/move-prover/tests/sources/functional/aborts_if.v2_exp [hunk 5]
```diff
@@ -102,7 +102,7 @@ error: abort not covered by any of the `aborts_if` clauses
     ┌─ tests/sources/functional/aborts_if.move:139:5
     │
 137 │           if (x == 2 || x == 3) abort 1;
-    │                         ------ abort happened here with code 0x1
```
[the remark this answers is not in the document]
Comment (rahxephon89) [reply third_party/move/move-prover/tests/sources/functional/aborts_if.v2_exp:105]: will file one once it is landed

File: third_party/move/move-model/bytecode/src/fat_loop.rs [hunk 6]
```diff
@@ -450,20 +450,57 @@ impl FatLoopBuilder {
...
                         .collect();
-                    block
+                    (*block_id, block)
                 },
             })
-            .collect()
+            .collect();
+        let mut results = vec![];
+        for (i, (block_id, block)) in blocks.iter().enumerate() {
+            // if this block has one successor and the last bc of this block is not a branch
```
Comment (wrwg) [inline third_party/move/move-model/bytecode/src/fat_loop.rs:474]: "If this is a fall-through block (block has one successor...)" we need to remove it by ...
The fatloop algorithm doesn't support fallthrough.

**Review by wrwg — APPROVED** [review]

**Pull request merged**